### ISSUE TITLE
undefined symbol: mysql_real_escape_string_quote

### ISSUE THREAD

**Issue by ostrokach** [PyMySQL/mysqlclient-python#110]
I am getting an error:

```
ImportError: /opt/conda/envs/_test/lib/python3.5/site-packages/_mysql.cpython-35m-x86_64-linux-gnu.so: undefined symbol: mysql_real_escape_string_quote
```

When trying to build a conda package with mysqlclient master branch. I think this is caused by commit #109?

**Comment by methane** [comment]
Thanks for using bleeding edge code and reporting issue.

I assume you use MySQL >= 5.7.4 when building extension module, and MySQL < 5.7.4 when running.

Could you try this?

```
$ mysql_config

$ ldd /opt/conda/envs/_test/lib/python3.5/site-packages/_mysql.cpython-35m-x86_64-linux-gnu.so
```

**Comment by ostrokach** [comment]
Thank you for your quick response!

```
$ ldd /opt/conda/envs/_test/lib/python3.5/site-packages/_mysql.cpython-35m-x86_64-linux-gnu.so

linux-vdso.so.1 =>  (0x00007fffb23f6000)
libmysqlclient.so.18 => /opt/conda/envs/_test/lib/python3.5/site-packages/../../libmysqlclient.so.18 (0x00007f3f980d7000)
libpthread.so.0 => /lib64/libpthread.so.0 (0x00007f3f97eb3000)
libm.so.6 => /lib64/libm.so.6 (0x00007f3f97c2f000)
librt.so.1 => /lib64/librt.so.1 (0x00007f3f97a27000)
libdl.so.2 => /lib64/libdl.so.2 (0x00007f3f97822000)
libpython3.5m.so.1.0 => /opt/conda/envs/_test/lib/python3.5/site-packages/../../libpython3.5m.so.1.0 (0x00007f3f972ff000)
libc.so.6 => /lib64/libc.so.6 (0x00007f3f96f6b000)
libstdc++.so.6 => /usr/lib64/libstdc++.so.6 (0x00007f3f96c64000)
libgcc_s.so.1 => /lib64/libgcc_s.so.1 (0x00007f3f96a4e000)
/lib64/ld-linux-x86-64.so.2 (0x000055a204b8f000)
libutil.so.1 => /lib64/libutil.so.1 (0x00007f3f9684b000
```

**Comment by ostrokach** [comment]
I get this error message when I try to compile `mysqlclient` using the conda-forge docker container: condaforge/linux-anvil:latest.

Conda recipe is as follows:

```
package:
  name: mysqlclient
  version: "1.3.8.dev"

source:
  git_url: https://github.com/PyMySQL/mysqlclient-python.git

build:
  number: 0
  skip: True # [osx]

requirements:
  build:
    - python
    - mysql-connector-c
  run:
    - python
    - mysql-connector-c

test:
  imports:
    - MySQLdb

about:
  home: https://github.com/PyMySQL/mysqlclient-python
  license: GNU General Public License (GPL)
  summary: 'Python interface to MySQL'
```

**Comment by methane** [comment]
What is `mysql_config` output?

About build error on conda-forge, have you read README?
https://github.com/PyMySQL/mysqlclient-python#prerequisites

If you have, please paste build error.

**Comment by ostrokach** [comment]
The reason I tried to use the most recent master of `mysqlclient` is that I started having some weird `Lost connection to MySQL server during query` errors when using `mysqlcleint 1.3.7` with `django 1.10.1`:

```
$ python manage.py runserver 8001
Performing system checks...

System check identified no issues (0 silenced).
Unhandled exception in thread started by <function check_errors.<locals>.wrapper at 0x7fc72a8776a8>
Traceback (most recent call last):
  File "/home/user/anaconda/envs/mum/lib/python3.5/site-packages/django/db/backends/utils.py", line 64, in execute
    return self.cursor.execute(sql, params)
  File "/home/user/anaconda/envs/mum/lib/python3.5/site-packages/django/db/backends/mysql/base.py", line 112, in execute
    return self.cursor.execute(query, args)
  File "/home/user/anaconda/envs/mum/lib/python3.5/site-packages/MySQLdb/cursors.py", line 228, in execute
    if not self._defer_warnings: self._warning_check()
  File "/home/user/anaconda/envs/mum/lib/python3.5/site-packages/MySQLdb/cursors.py", line 120, in _warning_check
    warnings = self._get_db().show_warnings()
  File "/home/user/anaconda/envs/mum/lib/python3.5/site-packages/MySQLdb/connections.py", line 368, in show_warnings
    self.query("SHOW WARNINGS")
  File "/home/user/anaconda/envs/mum/lib/python3.5/site-packages/MySQLdb/connections.py", line 280, in query
    _mysql.connection.query(self, query)
_mysql_exceptions.OperationalError: (2013, 'Lost connection to MySQL server during query')
```

My MySQL server version is 5.7.13:

```
$ mysql
Server version: 5.7.13-log MySQL Community Server (GPL)
```

**Comment by methane** [comment]
Stop writing unrelated logs.
"Lost connection to MySQL server" can be happen in many reasons.
And I'm not free tech support of MySQL.

**Comment by methane** [comment]
```
libmysqlclient.so.18 => /opt/conda/envs/_test/lib/python3.5/site-packages/../../libmysqlclient.so.18 (0x00007f3f980d7000)
```

I don't know anything about conda, but I suppose it provides own `libmysqlclient`.
And you may build mysqlclient-python with system libmysqlclient. It can cause such a
link error.

**Comment by ostrokach** [comment]
I think they could be related...

Anyway, `mysqlclient` README says that the dependencies are: `python-dev libmysqlclient-dev`.
These are the packages that are included in the conda build requirements.

`mysql-client-c` is a conda package containing `MySQL Connector C 6.1.6`: https://dev.mysql.com/downloads/connector/c/ (Linux - Generic (glibc 2.5) (x86, 64-bit), Compressed TAR Archive).

**Comment by methane** [comment]
> I think they could be related...

"Lost connection ..." error must not related.

**Comment by methane** [comment]
Again, I don't use conda and I don't know anything about conda.
Conda user forum must be better place to ask "how to build".

**Comment by ostrokach** [comment]
I have build many packages with conda before. This isn't the issue.

I was the person who contibuted the `mysqlclient 1.3.7` conda package to [bioconda](https://github.com/bioconda/bioconda-recipes/tree/b59318445d7fc225abf12f9bca641dcde293d6db/recipes/mysqlclient), which has [close to 3000](https://anaconda.org/search?q=mysqlclient) downloads to date.

I was just wondering what changes in mysqlclient master branch make the build fail, because I wa hoping that the master branch would fix my (possibly unrelated) problem...

**Comment by methane** [comment]
You didn't post mysql_config output
You didn't post build error log in docker

If you're expert, you must know about how mysql_config is important.

**Comment by ostrokach** [comment]
```
mysql_config 
Usage: /opt/conda/envs/_test/bin/mysql_config [OPTIONS]
Options:
        --cflags         [-I/opt/conda/envs/_test/include]
        --cxxflags       [-I/opt/conda/envs/_test/include]
        --include        [-I/opt/conda/envs/_test/include]
        --libs           [-L/opt/conda/envs/_test/lib -lmysqlclient -lpthread -lm -lrt -ldl]
        --libs_r         [-L/opt/conda/envs/_test/lib -lmysqlclient -lpthread -lm -lrt -ldl]
        --plugindir      [/usr/local/mysql/lib/plugin]
        --socket         [/tmp/mysql.sock]
        --port           [0]
        --version        [6.1.6]
        --variable=VAR   VAR is one of:
                pkgincludedir [/opt/conda/envs/_test/include]
                pkglibdir     [/opt/conda/envs/_test/lib]
                plugindir     [/usr/local/mysql/lib/plugin]
```

**Comment by methane** [comment]
OK, it may pointing right place.

What is output of `python setup.py build?`  In my env (Ubuntu):

```
$ python setup.py build_ext -if
...
gcc -pthread -shared -L/home/inada-n/pyenv/versions/2.7.11/lib build/temp.linux-x86_64-2.7/_mysql.o -L/usr/lib/x86_64-linux-gnu -lmysqlclient -lpthread -lz -lm -lrt -ldl -o build/lib.linux-x86_64-2.7/_mysql.so
copying build/lib.linux-x86_64-2.7/_mysql.so -> 

$ ldd _mysql.so
	linux-vdso.so.1 =>  (0x00007ffc04d2c000)
	libmysqlclient.so.20 => /usr/lib/x86_64-linux-gnu/libmysqlclient.so.20 (0x00007fb4888ff000)
...
```

**Comment by ostrokach** [comment]
```
$ python setup.py build
running build
running build_py
copying MySQLdb/release.py -> build/lib.linux-x86_64-3.5/MySQLdb
running build_ext
```

**Comment by ostrokach** [comment]
I put the conda build logs [here](https://gist.github.com/ostrokach/c3d7c04a795939186de3a1f76536cca9).

To me it looks like the `build_ext` output is similar to what you posted above...

**Comment by methane** [comment]
```
gcc -pthread -Wno-unused-result -Wsign-compare -DNDEBUG -g -fwrapv -O3 -Wall -Wstrict-prototypes -fPIC -Dversion_info=(1,3,7,'final',1) -D__version__=1.3.7 -I/opt/conda/envs/_build_placehold_placehold_placehold_placehold_placehold_placeho
/include -I/opt/conda/envs/_build_placehold_placehold_placehold_placehold_placehold_placeho/include/python3.5m -c _mysql.c -o build/temp.linux-x86_64-3.5/_mysql.o
In file included from /opt/conda/envs/_build_placehold_placehold_placehold_placehold_placehold_placeho/include/mysql.h:64,
                 from _mysql.c:30:
/opt/conda/envs/_build_placehold_placehold_placehold_placehold_placehold_placeho/include/mysql/client_plugin.h:97: warning: function declaration isn’t a prototype
In file included from /opt/conda/envs/_build_placehold_placehold_placehold_placehold_placehold_placeho/include/mysql.h:64,
                 from _mysql.c:30:
/opt/conda/envs/_build_placehold_placehold_placehold_placehold_placehold_placeho/include/mysql/client_plugin.h:107: warning: function declaration isn’t a prototype
_mysql.c: In function ‘_mysql_escape_string’:
_mysql.c:1068: warning: implicit declaration of function ‘mysql_real_escape_string_quote
```

It seems right header is used, but `mysql_real_escape_string_quote` is used. Strange...
What is your MYSQL_VERSION_ID in `/opt/conda/envs/.../include/mysql_version.h`?

**Comment by ostrokach** [comment]
```
$ cat /opt/conda/envs/_test/include/mysql_version.h
/* Copyright Abandoned 1996,1999 TCX DataKonsult AB & Monty Program KB
   & Detron HB, 1996, 1999-2004, 2007 MySQL AB.
   This file is public domain and comes with NO WARRANTY of any kind
*/

/* Version numbers for protocol & mysqld */

#ifndef _mysql_version_h
#define _mysql_version_h
#ifdef _CUSTOMCONFIG_
#include <custom_conf.h>
#else
#define PROTOCOL_VERSION            10
#define MYSQL_SERVER_VERSION       "5.7.6-m16"
#define MYSQL_VERSION_ID            50706
#define MYSQL_PORT                  3306
#define MYSQL_PORT_DEFAULT          0
#define MYSQL_UNIX_ADDR            "/tmp/mysql.sock"
#define MYSQL_CONFIG_NAME          "my"
#define MYSQL_COMPILATION_COMMENT  "MySQL Connector/C (GPL)"
#define LIBMYSQL_VERSION           "6.1.6"
#define LIBMYSQL_VERSION_ID         60106

/* mysqld compile time options */
#endif /* _CUSTOMCONFIG_ */

#ifndef LICENSE
#define LICENSE                     GPL
#endif /* LICENSE */

#endif /* _mysql_version_h */
```

**Comment by methane** [comment]
OK, I got!

MySQL 5.7.6 has new mysql_real_escape_string_quote() api.
https://dev.mysql.com/doc/relnotes/mysql/5.7/en/news-5-7-6.html#mysqld-5-7-6-feature

MySQL Connector/C 6.1.6 is released during cycle of 5.7.6, before the API is added.
So MYSQL_VERSION_ID in MySQL Connector/C 6.1.6 shows it's 5.7.6, but it doesn't
have the API added in MySQL 5.7.6.

Holy shit!
MySQL Connector/C release engineering seems totally broken.

**Comment by methane** [comment]
Fixed.  Thank you for your information.
